File: src/embroidermodder/select_box.py
#!/usr/bin/env python3

r"""
    Embroidermodder 2.

    ------------------------------------------------------------

    Copyright 2013-2022 The Embroidermodder Team
    Embroidermodder 2 is Open Source Software.
    See LICENSE for licensing terms.

    ------------------------------------------------------------

    The color selection docker class definition file.
"""

from libembroidery import Pen
from embroidermodder.utility import Brush, debug_message


class SelectBox():
    """
    The color selection docker.

    From a duplicated version:

    SelectBox(Shape s, QWidget* parent = 0)

    EmbColor leftBrushColor
    Color rightBrushColor
    Color leftPenColor
    Color rightPenColor
    unsigned char alpha

    QBrush dirBrush
    QBrush leftBrush
    QBrush rightBrush

    QPen dirPen
    QPen leftPen
    QPen rightPen

    int boxDir

    def setDirection(int dir)
    def setColors(colorL, fillL, colorR, fillR, newAlpha)
    def paintEvent(QPaintEvent*)
    def forceRepaint()
    def force_repaint():
        " . "
        #HACK: Take that QRubberBand!
        hack = size()
        resize(hack + QSize(1,1))
        resize(hack)
        # WARNING:
        # DO NOT SET THE QMDISUBWINDOW (this) FOCUSPROXY TO THE PROMPT
        # AS IT WILL CAUSE THE WINDOW MENU TO NOT SWITCH WINDOWS PROPERLY!
        # ALTHOUGH IT SEEMS THAT SETTING INTERNAL WIDGETS FOCUSPROXY IS OK.

        #    gview.setFocusProxy(self.mw.prompt);

        resize(sizeHint())

        curLayer = "0"
        curColor = 0; #TODO: color ByLayer
        curline_type = "ByLayer"
        curline_weight = "ByLayer"

        # Due to strange Qt4.2.3 feature the child window icon is not drawn
        # in the main menu if showMaximized() is called for a non-visible child window
        # Therefore calling show() first...
        show()
        showMaximized()

        setFocusPolicy(Qt_WheelFocus)
        setFocus()

        onWindowActivated()
    """
    def __init__(self, s, parent):
        " . "
        #Default values
        self.left_brush_color = "#FFFFFF"
        self.right_brush_color = "#FFFFFF"
        self.left_pen_color = "#FFFFFF"
        self.right_pen_color = "#FFFFFF"
        self.alpha = 1.0
        self.dir_brush = Brush()
        self.left_brush = Brush()
        self.right_brush = Brush()
        self.dir_pen = Pen()
        self.left_pen = Pen()
        self.right_pen = Pen()
        self.box_dir = 0
        self.set_colors("Qt_darkGreen", "Qt_green", "Qt_darkBlue", "Qt_blue", 32)

    def set_direction(self, dir):
        " . "
        if not dir:
            self.dir_pen = self.left_pen
            self.dir_brush = self.left_brush
        else:
            self.dir_pen = self.right_pen
            self.dir_brush = self.right_brush
        self.box_dir = dir

    def set_colors(self, color_left, fill_left, color_right, fill_right, new_alpha):
        " . "
        debug_message("SelectBox set_colors()")
        self.alpha = new_alpha

        self.left_pen_color = color_left
        # TODO: allow customization
        self.left_brush_color = fill_left + new_alpha
        self.right_pen_color = color_right
        # TODO: allow customization
        self.right_brush_color = fill_right + new_alpha

        self.left_pen.set_color(self.left_pen_color)
        self.left_pen.set_style("Qt_DashLine")
        self.left_brush.set_color(self.left_brush_color)
        self.left_brush.set_style("Qt_SolidPattern")

        self.right_pen.set_color(self.right_pen_color)
        self.right_pen.set_style("Qt_SolidLine")
        self.right_brush.set_color(self.right_brush_color)
        self.right_brush.set_style("Qt_SolidPattern")

        self.set_direction(self.box_dir)

        self.force_repaint()

    def paint_event(self, event):
        " . "
        painter = QPainter()
        painter.set_pen(self.dir_pen)
        painter.fill_rect(0, 0, width()-1, height()-1, self.dir_brush)
        painter.draw_rect(0, 0, width()-1, height()-1)

    def force_repaint(self):
        " . "
        #HACK: Take that QRubberBand!
        hack = size()
        resize(hack + (1, 1))
        resize(hack)

File: src/select_box.c
/* 
 *  Embroidermodder 2.
 *
 *  ------------------------------------------------------------
 *
 *  Copyright 2013-2022 The Embroidermodder Team
 *  Embroidermodder 2 is Open Source Software.
 *  See LICENSE for licensing terms.
 *
 *  ------------------------------------------------------------
 *
 *  Use Python's PEP7 style guide.
 *      https://peps.python.org/pep-0007/
 *
 *  Select box.
 */

#include "em2.h"

/* Default values.
 */
void
select_box_init(int tools)
{
    /*
    tools->left-brush.color = "#FFFFFF";
    tools->left-brush.style = BRUSH-STYLE-SOLID;
    tools->right-brush.color = "#FFFFFF";
    tools->right-brush.style = BRUSH-STYLE-SOLID;
    tools->left-pen.color = "#FFFFFF";
    tools->left-pen.style = LINE-STYLE-SOLID;
    tools->right-pen.color = "#FFFFFF";
    tools->right-pen.style = LINE-STYLE-SOLID;
    tools->alpha = 0xFF;
    tools->box-dir = 0;
    select-box-direction(tools, tools->box-dir);
    */
    /* not sure what this line was for */
    /*
    select_box_colors("darkGreen", "green", "darkBlue", "blue", 32);
    */
}

/* Choose between the left an right tools.
 */
void
select_box_direction(int tools, int dir)
{
    /*
    tools->box-dir = dir
    if (dir == DIRECTION-RIGHT) {
        tools->dir-pen = tools->right-pen
        tools->dir-brush = tools->right-brush
    }
    else {
        tools->dir-pen = tools->left-pen
        tools->dir-brush = tools->left-brush
    }
    */
}

void
set_view_select_box_colors(
    EmbColor colorL,
    EmbColor fillL,
    EmbColor colorR,
    EmbColor fillR,
    double alpha)
{
    printf("%d\n", colorL.r);
    printf("%d\n", fillL.r);
    printf("%d\n", colorR.r);
    printf("%d\n", fillR.r);
    printf("alpha %f\n", alpha);
     /* gview.setSelectBoxColors(colorL, fillL, colorR, fillR, alpha) */
}

/* TODO: allow customization. */
void
select_box_colors(Toolset *tools, char *setting)
{
    char setting_str[100];
    debug_message("SelectBox colors()");
    sprintf(setting_str, "%s-alpha", setting);
    tools->alpha = get_float(setting_str);

    sprintf(setting_str, "%s-color-left", setting);
    tools->left_pen.color = get_color(setting_str);
    tools->left_pen.style = get_int("line-style-dashed");

    sprintf(setting_str, "%s-fill-left", setting);
    tools->left_brush.color = get_color(setting_str);
    tools->left_brush.style = get_int("brush-style-solid");

    sprintf(setting_str, "%s-color-right", setting);
    tools->right_pen.color = get_color(setting_str);
    tools->right_pen.style = get_int("line-style-solid");

    sprintf(setting_str, "%s-fill-right", setting);
    tools->right_brush.color = get_color(setting_str);
    tools->right_brush.style = get_int("brush-style-solid");

    /*
    direction(tools->box_dir);

    select_box_force_repaint();
    */
}

/* Carry out a paint given the current pen and brush.
 * May need the canvas passing in.
 */
void
select_box_paint_event(int tools, int event)
{
    Painter *painter = create_painter();
    /*
    painter->pen->dir_pen = 0;
    painter.fill-rect(0, 0, width()-1, height()-1, dir-brush);
    painter.draw-rect(0, 0, width()-1, height()-1);
    */
}

/* HACK: Take that QRubberBand! */
void
select_box_force_repaint()
{
    /*
    hack = size();
    resize(hack + (1, 1));
    resize(hack);
    */
}

